File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.java
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        if(nums1.length > nums2.length) return intersection(nums2, nums1);
        
        Set<Integer> set = new HashSet<>();
        List<Integer> ans = new ArrayList<>();
        
        for(int x : nums1) {
            set.add(x);
        }
        for(int x : nums2) {
            if(set.contains(x)) {
                ans.add(x);
                set.remove(x);
            }
        }
        
        int[] ls = new int[ans.size()];
        for(int i = 0; i < ans.size(); i++) {
            ls[i] = ans.get(i);
        }
        return ls;
    }
}

File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.ts
type Obj = {[key: number]: number};
const reducerhelper = (nums: number[]): Obj => {
    let map1: Obj  = nums.reduce((acc: Obj, num: number) => {
        if(!acc[num]) acc[num] = 0;
        acc[num]++;
        return acc;
    }, {}); 
    return map1;
}
function intersection(nums1: number[], nums2: number[]): number[] {
    const ans: number[] = [];
    let map1: Obj  = reducerhelper(nums1);
    let map2: Obj  = reducerhelper(nums2);
    
    for(let [key, val] of Object.entries(map1)) {
        if(map2[key]) ans.push(Number(key));
    }
    return ans;
};